contracts/staking/ServiceStakingToken.sol

```diff
@@ -0,0 +1,199 @@
+    function mapServiceIdTokenDeposit(uint256 serviceId) external view returns (address, uint96);
+}
+
+/// @dev Failure of a transfer.
+/// @param token Address of a token.
+/// @param from Address `from`.
+/// @param to Address `to`.
+/// @param value Value.
+error TransferFailed(address token, address from, address to, uint256 value);
+
+/// @title ServiceStakingToken - Smart contract for staking the service by its owner having an ERC20 token as a deposit
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+/// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
+contract ServiceStakingToken is ServiceStakingBase {
+    // ServiceRegistryTokenUtility address
+    address public immutable serviceRegistryTokenUtility;
+    // Security token address for staking corresponding to the service deposit token
+    address public immutable securityToken;
+
+    /// @dev ServiceStakingToken constructor.
+    /// @param _apy Staking APY (in single digits).
+    /// @param _minSecurityDeposit Minimum security deposit for a service to be eligible to stake.
+    /// @param _stakingRatio Staking ratio: number of seconds per nonce (in 18 digits).
+    /// @param _serviceRegistry ServiceRegistry contract address.
+    /// @param _serviceRegistryTokenUtility ServiceRegistryTokenUtility contract address.
+    /// @param _securityToken Address of a service security token.
+    constructor(
+        uint256 _apy,
+        uint256 _minSecurityDeposit,
+        uint256 _stakingRatio,
+        address _serviceRegistry,
+        address _serviceRegistryTokenUtility,
+        address _securityToken
+    )
+        ServiceStakingBase(_apy, _minSecurityDeposit, _stakingRatio, _serviceRegistry)
+    {
+        // TODO: calculate minBalance
+        // Initial checks
+        if (_securityToken == address(0) || _serviceRegistryTokenUtility == address(0)) {
+            revert ZeroAddress();
+        }
+
+        securityToken = _securityToken;
+        serviceRegistryTokenUtility = _serviceRegistryTokenUtility;
+    }
+
+    /// @dev Safe token transferFrom implementation.
+    /// @notice The implementation is fully copied from the audited MIT-licensed solmate code repository:
+    ///         https://github.com/transmissions11/solmate/blob/v7/src/utils/SafeTransferLib.sol
+    ///         The original library imports the `ERC20` abstract token contract, and thus embeds all that contract
+    ///         related code that is not needed. In this version, `ERC20` is swapped with the `address` representation.
+    ///         Also, the final `require` statement is modified with this contract own `revert` statement.
+    /// @param token Token address.
+    /// @param from Address to transfer tokens from.
+    /// @param to Address to transfer tokens to.
+    /// @param amount Token amount.
+    function safeTransferFrom(address token, address from, address to, uint256 amount) internal {
```

Comment:
Why does this need to be copied here?

Reply:
It is provided as a library in solmate, however it pulls the full ERC20 token contract for that, and we don't want that.

Just figured it's not needed at all as our secured tokens are already checked for "safe" when depositing the first time in the ServiceRegistryTokenUtility contract.